File: leetcode/Algorithms/733.FloodFill/Solution.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        boolean[][] visited = new boolean[image.length][image[0].length];
        int origin = image[sr][sc];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{sr, sc});
        while (!queue.isEmpty()) {
            int tmp[] = queue.poll(), x = tmp[0], y = tmp[1];
            if (visited[x][y]) continue;
            visited[x][y] = true;
            image[x][y] = newColor;
            if (x != 0 && !visited[x - 1][y] && image[x - 1][y] == origin) queue.add(new int[]{x - 1, y});
            if (y != 0 && !visited[x][y - 1] && image[x][y - 1] == origin) queue.add(new int[]{x, y - 1});
            if (x != image.length - 1 && !visited[x + 1][y] && image[x + 1][y] == origin)
                queue.add(new int[]{x + 1, y});
            if (y != image[0].length - 1 && !visited[x][y + 1] && image[x][y + 1] == origin)
                queue.add(new int[]{x, y + 1});
        }
        return image;
    }
}

File: leetcode/Algorithms/733.FloodFill/Solution.py
class Solution:
    def floodFill(self, image, sr, sc, newColor):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type newColor: int
        :rtype: List[List[int]]
        """
        if image[sr][sc] != newColor:
            self.helper(image, sr, sc, newColor, image[sr][sc])
        return image

    def helper(self, image, sr, sc, newColor, origin):
        image[sr][sc] = newColor
        if sr != 0 and image[sr - 1][sc] == origin:
            self.helper(image, sr - 1, sc, newColor, origin)
        if sc != 0 and image[sr][sc - 1] == origin:
            self.helper(image, sr, sc - 1, newColor, origin)
        if sr != len(image) - 1 and image[sr + 1][sc] == origin:
            self.helper(image, sr + 1, sc, newColor, origin)
        if sc != len(image[0]) - 1 and image[sr][sc + 1] == origin:
            self.helper(image, sr, sc + 1, newColor, origin)

